README.md
[![Build Status](https://travis-ci.com/TheSpicyMeatball/time-input-formatter.svg?branch=main)](https://travis-ci.com/TheSpicyMeatball/time-input-formatter)
[![Coverage Status](https://coveralls.io/repos/github/TheSpicyMeatball/time-input-formatter/badge.svg?branch=main)](https://coveralls.io/github/TheSpicyMeatball/time-input-formatter?branch=main)
[![dependencies Status](https://status.david-dm.org/gh/TheSpicyMeatball/time-input-formatter.svg?path=dist)](https://david-dm.org/TheSpicyMeatball/time-input-formatter)
[![npm version](https://badge.fury.io/js/time-input-formatter.svg)](https://badge.fury.io/js/time-input-formatter)

# time-input-formatter

<p>Interpret and format a user inputted time string or partial time string to 12hr or 24hr time</p>

> This library contains es5 and es6 implementations in the deliverable and is fully tree-shakable for es6.

<p><b>Version:</b> 1.0.0</p>


  <h2>deriveTimeFormat</h2>
<p>Derive a time format from a user inputted string</p>
<p>Since v1.0.0</p>
<table>
      <thead>
      <tr>
        <th>Param</th>
        <th>Type</th></tr>
      </thead>
      <tbody><tr><td><p><b>value</b></p>The input value to derive the format</td><td>string</td></tr><tr><td><p><b>format</b></p>The desired output format</td><td>Format</td></tr></tbody>
    </table><p><b>Returns:</b> {ParsedTime}</p><h4>Supporting Types</h4>

```
export type Format = '12hm' | '12hms' | '24hm' | '24hms';
export type ParsedTime = { valid: boolean, value: string };
```
<h4>Import</h4>

```
import { deriveTimeFormat, Format, ParsedTime } from 'time-input-formatter';
```

  


## Example Inputs &amp; Outputs

<p>These are a few sample partial inputs and how they will be interpreted by the util. These are just a few of the many that it can handle.</p>
<table>
  <thead>
    <tr>
      <th>Input</th>
      <th>12hm Output</th>
      <th>12hms Output</th>
      <th>24hm Output</th>
      <th>24hms Output</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code>7</code></td>
      <td><code>7:00 AM</code></td>
      <td><code>7:00:00 AM</code></td>
      <td><code>07:00</code></td>
      <td><code>07:00:00</code></td>
    </tr>
    <tr>
      <td><code>7a</code></td>
      <td><code>7:00 AM</code></td>
      <td><code>7:00:00 AM</code></td>
      <td><code>07:00</code></td>
      <td><code>07:00:00</code></td>
    </tr>
    <tr>
      <td><code>7am</code></td>
      <td><code>7:00 AM</code></td>
      <td><code>7:00:00 AM</code></td>
      <td><code>07:00</code></td>
      <td><code>07:00:00</code></td>
    </tr>
    <tr>
      <td><code>7p</code></td>
      <td><code>7:00 PM</code></td>
      <td><code>7:00:00 PM</code></td>
      <td><code>19:00</code></td>
      <td><code>19:00:00</code></td>
    </tr>
    <tr>
      <td><code>700</code></td>
      <td><code>7:00 AM</code></td>
      <td><code>7:00:00 AM</code></td>
      <td><code>07:00</code></td>
      <td><code>07:00:00</code></td>
    </tr>
    <tr>
      <td><code>0700</code></td>
      <td><code>7:00 AM</code></td>
      <td><code>7:00:00 AM</code></td>
      <td><code>07:00</code></td>
      <td><code>07:00:00</code></td>
    </tr>
    <tr>
      <td><code>715</code></td>
      <td><code>7:15 AM</code></td>
      <td><code>7:15:00 AM</code></td>
      <td><code>07:15</code></td>
      <td><code>07:15:00</code></td>
    </tr>
    <tr>
      <td><code>71527</code></td>
      <td><code>7:15 AM</code></td>
      <td><code>7:15:27 AM</code></td>
      <td><code>07:15</code></td>
      <td><code>07:15:27</code></td>
    </tr>
    <tr>
      <td><code>16</code></td>
      <td><code>4:00 PM</code></td>
      <td><code>4:00:00 PM</code></td>
      <td><code>16:00</code></td>
      <td><code>16:00:00</code></td>
    </tr>
    <tr>
      <td><code>0</code></td>
      <td><code>12:00 AM</code></td>
      <td><code>12:00:00 AM</code></td>
      <td><code>00:00</code></td>
      <td><code>00:00:00</code></td>
    </tr>
    <tr>
      <td><code>27</code></td>
      <td><code>12:00 AM</code></td>
      <td><code>12:00:00 AM</code></td>
      <td><code>00:00</code></td>
      <td><code>00:00:00</code></td>
    </tr>
  </tbody>
</table>

<a href="#license"></a>
<h2>License</h2>

MIT


<a href="#author"></a>
<h2>Author</h2>
Michael Paravano




<a href="#dependencies"></a>
<h2>Dependencies</h2>



<table>

  <tr>
    <td>
      @paravano/utils
    </td>
    <td>
      ^1.4.0
    </td>
  </tr>

  <tr>
    <td>
      date-fns
    </td>
    <td>
      ^2.21.1
    </td>
  </tr>

</table>


CHANGELOG.md
## [1.0.0] - 2021-04-23
- Initial release

src/utils/index.ts
import { parse, isValid, format as formatDate } from 'date-fns';
import { first, isNilOrEmpty, last, or, pluck, replaceAt, take } from '@paravano/utils';
import { Format, Format12Hr, Format24Hr, ParsedTime } from '../types';

const meridian = (value: string) : string => value.includes('p') ? 'PM' : 'AM';
const minutes = (value: string) : string => parseInt(value.slice(0, 2), 10) < 60 ? value.slice(0, 2) : '00';
const seconds = (value: string) : string => parseInt(value.slice(0, 2), 10) < 60 ? value.slice(0, 2) : '00';

export const timeFormats = {
  _12hm: 'h:mm a',
  _12hms: 'h:mm:ss a',
  _24hm: 'HH:mm',
  _24hrs: 'HH:mm:ss',
};

/**
 * Try to interpret partial entry for 12hr time
 *
 * @param {string} value - The input value to derive the format
 * @param {Format12Hr} format - The desired output format
 * @returns {ParsedTime}
 */
 export const try12Hr = (value: string, format: Format12Hr) : ParsedTime => {
  if (isNilOrEmpty(value)) return;

  const hour = (value: string) : string => {
    const int = parseInt(value, 10);

    if (int < 10 || or(int, 10, 11, 12)) return int.toString();
    if (int >= 24) return '12';
    
    return (int - 12).toString();
  };

  // 0 | 00000000 | 0a | 0A | 0am | 0p | 0P | 0pm | 0PM => 12:00 AM
  if (/^0*(a|am|p|pm)?$/i.test(value)) {
    return { valid: true, value: `12:00${format === timeFormats._12hms ? ':00' : ''} AM` };
  }

  // remove leading zeros
  const _value = value.replace(/^0*/, '').toLowerCase();

  // 7 | 7a | 7am | 7A | 7AM   => 7:00 AM
  // 7p | 7pm | 7P | 7PM            => 7:00 PM
  if (/^[0-9](a|am|p|pm)?$/.test(_value)) {
    const num = _value.replace(/a|p|m/g, '');
    return { valid: true, value: `${num}:00${format === timeFormats._12hms ? ':00' : ''} ${meridian(_value)}` };
  }

  // 10 | 10a | 10am | 10A | 10AM  => 10:00 AM
  // 12 | >= 24  => 12:00 AM
  // 16          => 4:00 PM
  if (/^(1|2)[0-9]{3,}(a|am|p|pm)?$/.test(_value) || /^(1|2)[0-9]{1}(a|am|p|pm)?$/.test(_value)) {
    const num = take(Array.from(_value.replace(/a|am|p|pm$/g, '')), format === timeFormats._12hms ? 6 : 4);
    const h = hour(take(num, 2).join(''));
    const m = minutes(pluck(num, [2, 3]).map(x => x === undefined ? '0' : x).join(''));
    const s = format === timeFormats._12hms ? ':' + seconds(pluck(num, [4, 5]).map(x => x === undefined ? '0' : x).join('')) : '';

    if (parseInt(take(num, 2).join('')) <= 12) return { valid: true, value: `${h}:${m}${s} ${meridian(_value)}` };
    if (parseInt(take(num, 2).join('')) >= 24) return { valid: true, value: `${h}:${m}${s} AM` };

    return { valid: true, value: `${h}:${m}${s} PM` };
  }

  // 700 | 0700 | 799 | 77777777  => 7:00 AM
  // 715                          => 7:15 AM
  if (/[0-9]{3}[0-9]*(a|am|p|pm)?$/.test(_value)) {
    const num = take(Array.from(_value.replace(/a|am|p|pm$/g, '')), format === timeFormats._12hms ? 5 : 3);
    return { valid: true, value: `${first(num, '0')}:${minutes(pluck(num, [1, 2]).join(''))}${format === timeFormats._12hms ? ':' + seconds(pluck(num, [3, 4]).join('')) : ''} ${meridian(_value)}` };
  }
};

/**
 * Try to interpret partial entry for 24hr time
 *
 * @param {string} value - The input value to derive the format
 * @param {Format24Hr} format - The desired output format
 * @returns {ParsedTime}
 */
export const try24Hr = (value: string, format: Format24Hr) : ParsedTime => {
  if (isNilOrEmpty(value)) return;

  const _value = value.toLowerCase();

  const hour = (value: string) : string => {
    const _meridian = meridian(_value);
    let int = parseInt(value.replace(/^0*|a|am|p|pm$/g, ''), 10);

    int = isNaN(int) ? 0 : int;

    if (int < 12) {
      if (_meridian === 'PM') {
        return (int + 12).toString();
      }

      if (int < 10) {
        return '0' + int.toString();
      }

      return int.toString();
    }

    if (int >= 24 || (int === 12 && value.includes('a'))) return '00';

    return int.toString();
  };

  // 0 | 0a | 0am | 0A | 0AM | 0p | 0pm | 0P | 0PM | 25:00  => 00:00
  if (/^0+(a|am|p|pm)?$/.test(_value) || /^2[4-9]:[0-9]{2}$/.test(_value)) {
    return { valid: true, value: `00:00${format === timeFormats._24hrs ? ':00' : ''}` };
  }

  // 7 | 07 | 7a | 7am | 7A | 7AM   => 07:00
  // 7p | 7pm | 7P | 7PM            => 19:00
  if (/^0*[0-9](a|am|p|pm)?$/.test(_value)) {
    return { valid: true, value: `${hour(_value)}:00${format === timeFormats._24hrs ? ':00' : ''}` };
  }

  // 10 | 10a | 10am | 10A | 10AM  => 10:00
  // >= 24                         => 00:00
  // 16                            => 16:00
  if (/^(1|2)[0-9]{3,}(a|am|p|pm)?$/.test(_value) || /^(1|2)[0-9]{1}(a|am|p|pm)?$/.test(_value)) {
    const num = take(Array.from(_value.replace(/^0|a|am|p|pm$/g, '')), format === timeFormats._24hrs ? 6 : 4);
    const _meridian = _value.includes('a') || _value.includes('p') ? meridian(_value).toLowerCase() : '';
    const h = hour(take(num, 2).join('') + _meridian);
    const m = minutes(pluck(num, [2, 3]).map(x => x === undefined ? '0' : x).join(''));
    const s = format === timeFormats._24hrs ? ':' + seconds(pluck(num, [4, 5]).map(x => x === undefined ? '0' : x).join('')) : '';

    if (parseInt(take(num, 2).join('')) <= 12) return { valid: true, value: `${h}:${m}${s}` };
    if (parseInt(take(num, 2).join('')) >= 24) return { valid: true, value: `00:00${format === timeFormats._24hrs ? ':00' : ''}` };

    return { valid: true, value: `${h}:${m}${s}` };
  }

  // 700 | 0700 | 799 | 77777777  => 7:00
  // 715                          => 7:15
  if (/^0?[0-9]{3}[0-9]*(a|am|p|pm)?$/.test(_value)) {
    let num = take(Array.from(_value.replace(/^0|a|am|p|pm$/g, '')), format === timeFormats._24hrs ? 5 : 3);
    
    if (parseInt(num.join(''), 10) === 0) return { valid: true, value: `00:00${format === timeFormats._24hrs ? ':00' : ''}` };

    if (meridian(_value) === 'PM') {
      num = replaceAt(num, 0, (parseInt(first(num, '0'), 10) + 12).toString());
    } else {
      num = replaceAt(num, 0, '0' + first(num, '0'));
    }

    return { valid: true, value: `${first(num, '00')}:${minutes(pluck(num, [1, 2]).join(''))}${format === timeFormats._24hrs ? ':' + seconds(pluck(num, [3, 4]).join('')) : ''}` };
  }

  // 7:00000000000 | 7:00000000001 | 07:00000000000 | 07:00000000001 => 07:00
  if (/^[0-9]{1,2}:[0-9]*(a|am|p|pm)?$/.test(_value)) {
    const ary = _value.split(':');
    const h = hour(first(ary, '00'));
    const m = minutes(last(ary, '00').replace(/am|pm|a|p/i, ''));
    const s = format === timeFormats._24hrs ? ':' + seconds(pluck(Array.from(last(ary, '00').replace(/am|pm|a|p/i, '')), [2, 3]).map(x => x === undefined ? '0' : x).join('')) : '';

    return { valid: true, value: `${h}:${m}${s}` };
  }
};

/**
 * If no partial entry patter was found, try to parse with this fallback
 *
 * @param {string} value - The input value to derive the format
 * @param {Format} format - The desired output format
 * @returns {ParsedTime}
 */
export const fallback = (value: string, format: Format) : ParsedTime => {
  if (/^[0-9]+:?[0-9]+(a|am|p|pm)$/i.test(value)) {
    value = value.replace(/(a|am|p|pm)$/i, ' $1');
  }
  const time = parse(value, format, new Date());

  if (isValid(parse(value, format, new Date()))) {
    return {
      valid: true,
      value: formatDate(time, format),
    };
  }

  // try each time format
  const keys = Object.keys(timeFormats);

  for (let i = 0; i < keys.length; i++) {
    const _format = timeFormats[keys[i]];
    const time = parse(value, _format, new Date());

    if (isValid(time)) {
      return {
        valid: true,
        value: formatDate(time, format),
      };
    }
  }

  return { valid: false, value };
};